### ISSUE TITLE
change activity item type handing due to v2 backend changes

### ISSUE THREAD

**Issue by petervojtek** [PowerlineApp/powerline-mobile#187]
currently it is not possible to view activity item detail as yesteday `GET /v2/activities` has altered. also create new user content is broken.

```
Igor: 

we've refactored micropetitions.
we split it to `post` and `user-petition`

old endpoints must work correctly, but only `user-petition` will be created, attribute `type` is ignored. you can check new docs, sections `Posts` and `User Petitions`
there are some changes with answer
```

**Comment by petervojtek** [comment]
@igaponov , when I try to access `user post` details which endpoint am I supposed to use?

the `User post` (ID=309) was created before the last api changes.

```
Request URL:https://api-dev.powerli.ne/api/micro-petitions/309
Request Method:GET
Status Code:500 Internal Server Error

{"code":500,"message":"Class \"Civix\\CoreBundle\\Entity\\Micropetitions\\Petition\" is not a valid entity or mapped super class.","errors":null}
```

I also tried this endpoint:

```
Request URL:https://api-dev.powerli.ne/api/v2/posts/309
Request Method:GET
Status Code:404 Not Found

{"code":404,"message":"Civix\\CoreBundle\\Entity\\Post object not found.","errors":null}
```

and also this one:

```
Request URL:https://api-dev.powerli.ne/api/v2/user-petitions/309
Request Method:GET
Status Code:404 Not Found

{"code":404,"message":"Civix\\CoreBundle\\Entity\\UserPetition object not found.","errors":null}
```

UPDATE:

now I see, that  it is not found because this is how `activity.entity` looks like: `{"type":"user-petition","id":null,"group_id":285}`. Which ID am I supposed to use and is it intentional that `activity.entity.id` is null?

```
{"zone":"expired","id":501,"title":"Desalinate pacific ocean","description":"test.","description_html":"","sent_at":"2016-08-05T08:09:20.000Z","expire_at":"2016-08-06T07:47:16.000Z","responses_count":1,"owner":{"id":73,"type":"user","official_title":"","first_name":"Peter10","last_name":"Peter10","avatar_file_path":"http://devpowerline.s3-website-eu-west-1.amazonaws.com/avatars/577cb68e6f276.jpeg"},"entity":{"type":"user-petition","id":null,"group_id":285},"picture":null,"read":false,"rate_up":null,"rate_down":null,"image_src":null,"updated_at":"2016-08-05T08:09:20+0000","petition":null,"post":null,"metadata":null,"comments_count":0,"answers":[],"quorum":1,"owner_info_1":"San Diego"}
```

**Comment by igaponov** [comment]
@petervojtek fixed. Activity with id=501 has a user petition with id=237.

**Comment by petervojtek** [comment]
@jterps08 , @austinpapp , I just want to make sure you are aware of the consequences of [the changes made by Igor above](https://github.com/PowerlineApp/powerline-mobile/issues/187#issue-172059335) on backend -- it will take several man-days to adapt to the changes on mobile app side.

The reason is that before the changes were made, all the user-content was handled via petition business logic -- which is not good, but at least the mobile app was already adapted to that.

Now the backend has split the user content handling to use separate api endpoint for each content type (user post, user petition, ..). This means that I have to take the petition controller (and html) which is now used for user-post, user-petition and petition, and split it into three controllers and html templates and adapt it to each of the appropriate content types. Also the attributes of the content types seem to have changed to some extent which I will have to adapt to.

I actually welcome the changes made by Igor on backend, however I am just not sure if you are aware of the time required to adapt to this changes in mobile app.

**Comment by jterps08** [comment]
@petervojtek I was afraid of this. We need to make sure you and Igor discuss any re-architecture/refactoring efforts moving forward to understand full impact of issues and that these impacts are communicated to us so that we can make a go/no-go decision. Here's my question for you, Peter: You say 'several man days'. Can we complete this and the other P1/P2 bugs / P1 enhancements before Sept 2 or only without this? And... I am guessing we need to implement this change before we start doing the web app. Thoughts?

@austinpapp This is not required for going live. Can we back out this change and re-instate it later after go-live for the 2.1 release? Assuming that Peter can't fit this in between now and 9/2, we don't have time for this. P1/P2 bugs and P1 enhancements take priority over refactoring efforts.

**Comment by petervojtek** [comment]
@jterps08 , my time estimate is based on today's experience with  [issue #122 which's goal was to adapt to changes made to followers in v2 api](https://github.com/PowerlineApp/powerline-mobile/issues/122#issuecomment-240643555) -- my initial estimate was 3-4 hours, at the end I spent with it 8 hours and there are still some outstanding issues but principally it is done.

This issue (adapting to changes to user content types) seems to be like 2x larger than #122.

**Comment by jterps08** [comment]
Okay, thanks @petervojtek . So #187 could be about 16 hours... That's almost a full week's worth of work. How do we feel about the level of effort for the other P1/P2 bugs and P1 enhancements? 

Right now, it seems like we need to take the following approach:
1. P1/P2 bugs and P1 enhancements (with notifications testing)
2. Production Release
3. 187 work
4. Web app work

Austin?

**Comment by petervojtek** [comment]
update: will be resolved in #191

@igaponov , I try to render this user-petition detail:

```
GET https://api-dev.powerli.ne/api/v2/user-petitions/237

{"quorum_count":1,"share_picture":"http://devpowerline.s3-website-eu-west-1.amazonaws.com/avatars/577cb68e6f276.jpeg","petition_body_html":"","petition_body":"test.","expire_at":"2017-08-23T15:19:44+0000","user_expire_interval":0,"type":"long petition","link":"","publish_status":0,"answers":[{"option_id":1,"id":10,"user":{"full_name":"Peter21 Peter21","id":77,"type":"user","username":"Peter21","first_name":"Peter21","last_name":"Peter21","official_title":"Peter21 Peter21","email":"peter21@test.com","zip":"12","avatar_file_name":"https://api-dev.powerli.ne/bundles/civixfront/img/default_user.png","birth":"01/01/1977","address1":"San Diego","address2":"21","city":"San Diego","state":"CA","country":"US","phone":"ds","facebook_link":null,"twitter_link":null,"race":null,"sex":null,"orientation":null,"marital_status":null,"religion":null,"employment_status":null,"income_level":null,"education_level":null,"party":null,"philosophy":null,"donor":null,"bio":null,"slogan":null,"interests":[],"registration":null,"token":"36isoe2ud1icowk4sckgokwk804c8kwoc8ooskwo08884gckk77","facebook_id":null,"ios_device":null,"android_device":null,"do_not_disturb":false,"is_notif_questions":true,"is_notif_discussions":true,"is_notif_messages":true,"is_notif_micro_following":true,"is_notif_micro_group":true,"is_notif_scheduled":false,"is_notif_own_post_changed":true,"scheduled_from":null,"scheduled_to":null,"is_registration_complete":true,"plain_password":null},"created_at":"2016-08-05T08:09:20+0000"}],"is_subscribed":false,"id":237,"title":"Desalinate pacific ocean","body":"test.","html_body":"","is_outsiders_sign":false,"created_at":"Fri, 05 Aug 2016 07:47:16 +0000","user":{"full_name":"Peter10 Peter10","id":73,"type":"user","username":"Peter10","first_name":"Peter10","last_name":"Peter10","official_title":"Peter10 Peter10","email":"peter10@test.com","zip":"123","avatar_file_name":"http://devpowerline.s3-website-eu-west-1.amazonaws.com/avatars/577cb68e6f276.jpeg","birth":"01/01/1977","address1":"San Diego","address2":"123","city":"San Diego","state":"CA","country":"US","phone":"+1123","facebook_link":null,"twitter_link":null,"race":null,"sex":null,"orientation":null,"marital_status":null,"religion":null,"employment_status":null,"income_level":null,"education_level":null,"party":null,"philosophy":null,"donor":null,"bio":"whatever bio be heree","slogan":"live and let die","interests":[],"registration":null,"token":"5942cd6q6look4wgww8go0gkkk0088sccoo08cc8ko84wwk84073","facebook_id":null,"ios_device":null,"android_device":null,"do_not_disturb":false,"is_notif_questions":true,"is_notif_discussions":true,"is_notif_messages":true,"is_notif_micro_following":true,"is_notif_micro_group":true,"is_notif_scheduled":false,"is_notif_own_post_changed":true,"scheduled_from":null,"scheduled_to":null,"is_registration_complete":true,"plain_password":null},"boosted":false,"organization_needed":false,"signatures":[{"option_id":1,"id":10,"user":{"full_name":"Peter21 Peter21","id":77,"type":"user","username":"Peter21","first_name":"Peter21","last_name":"Peter21","official_title":"Peter21 Peter21","email":"peter21@test.com","zip":"12","avatar_file_name":"https://api-dev.powerli.ne/bundles/civixfront/img/default_user.png","birth":"01/01/1977","address1":"San Diego","address2":"21","city":"San Diego","state":"CA","country":"US","phone":"ds","facebook_link":null,"twitter_link":null,"race":null,"sex":null,"orientation":null,"marital_status":null,"religion":null,"employment_status":null,"income_level":null,"education_level":null,"party":null,"philosophy":null,"donor":null,"bio":null,"slogan":null,"interests":[],"registration":null,"token":"36isoe2ud1icowk4sckgokwk804c8kwoc8ooskwo08884gckk77","facebook_id":null,"ios_device":null,"android_device":null,"do_not_disturb":false,"is_notif_questions":true,"is_notif_discussions":true,"is_notif_messages":true,"is_notif_micro_following":true,"is_notif_micro_group":true,"is_notif_scheduled":false,"is_notif_own_post_changed":true,"scheduled_from":null,"scheduled_to":null,"is_registration_complete":true,"plain_password":null},"created_at":"2016-08-05T08:09:20+0000"}],"cached_hash_tags":[],"metadata":null,"owner":{"id":73,"avatar":"http://devpowerline.s3-website-eu-west-1.amazonaws.com/avatars/577cb68e6f276.jpeg","type":"user","name":"Peter10 Peter10"}}
```

and the existing code is build around processing and displying some attribute named `options`. in the [legacy business logic](https://github.com/PowerlineApp/powerline-mobile/blob/develop/www/js/services/utils.js#L102) it seems the `options` array provided information on how the users voted per each petition option/answer. 

But now there is no `options` attribute provided by `GET https://api-dev.powerli.ne/api/v2/user-petitions/237` and I am not sure how should I proceed -- am I supposed to process `answers` attribute to render how users cumulatively voted for each answer type?

**Comment by petervojtek** [comment]
update: will be resolved in #190.

@igaponov , it seems the user-petition comments [do not have v2 endpoint](https://api-dev.powerli.ne/api-doc#get--api-{typeEntity}-{entityId}-comments-). Am I supposed to use `https://api-dev.powerli.ne/api/micro-petitions/{user-petition-id}/comments/` for the purpose of CRUD manipulation of user petition comments?

**Comment by petervojtek** [comment]
update - will be resolved in #189.

@igaponov, when I try to upvote a post  (created successfully via the latest v2 api), I experience this:

```
Request URL:https://api-dev.powerli.ne/api/v2/posts/240/vote
Request Method:POST
Status Code:500 Internal Server Error

{"option":"upvote"}

response:
{"code":500,"message":"An exception occurred while executing 'INSERT INTO post_votes (option, created_at, user_id, post_id) VALUES (?, ?, ?, ?)' with params [1, \"2016-08-24 16:48:12\", 75, 240]:\n\nSQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'option, created_at, user_id, post_id) VALUES (1, '2016-08-24 16:48:12', 75, 240)' at line 1","errors":null}
```

**Comment by petervojtek** [comment]
ok, this issue is too messy and the main job was done.
 I will split it into several more focused issues.
